### DISCUSSION TITLE
I want to create a console client of my own, hoping to receive advice

### DISCUSSION THREAD

**Discussion by iTsThermo** [ORelio/Minecraft-Console-Client#2777] [General]
Hi, I'm a university student majoring in Computer Science, minor in Cyber Security. I'm familiar with programming and am currently interning as a Software Developer. I want to learn more about the methodology behind creating a console client (I'm going into this blind since my research was a complete failure, I even saw ORelio talking about CC's from 2011) and I would like to create my own from scratch (a barebones version) to use on a server that I play on with my friends. I played Minecraft factions before so I'm familiar with console clients, but I'm not familiar with what's under the hood. It would be an awesome project to put on my resume and learning opportunity for me. Here is my e-mail: vansh.kanojia@yahoo.com , feel free to contact me.

**Comment by breadbyte** [comment]
A place where all the packets Minecraft sends and receives are documented and can be found at [wiki.vg](https://wiki.vg/Main_Page).

To start with a basic gist of how connecting to a server works "under the hood", here's a general process of how to login to a server.

There are three "phases" in logging in to a Minecraft server.

- The initial phase is the "Login" phase, where the client communicates details to the server regarding the login procedure.

- The second phase is the "Configuration" phase, which shares information between the client and the server about their capabilities (such as the current server draw distance, or which data packs are installed in both the client and the server.)

- The third phase is the "Play" phase, in which a physical player is then spawned on the server.

The back and forth with the server, documented in the [Protocol FAQ in wiki.vg](https://wiki.vg/Protocol_FAQ), is as follows:
1. Open a TCP network connection to the server.
2. Send a [Handshake Packet](https://wiki.vg/Protocol#Handshake) to the server. 
This is us saying "Hello, I want to go to the login phase" to the server.

The server and the client will now move into the "Login" phase.

3. Send a [Login Start Packet](https://wiki.vg/Protocol#Login_Start) to the server. 
This is us saying "I want to log in as this player" to the server.

4. The server will return an [Encryption Request Packet](https://wiki.vg/Protocol#Encryption_Request) (if it is a "premium" server).
The client and server will then share the encryption key and secret.

5. The server will send a [Set Compression Packet](https://wiki.vg/Protocol#Set_Compression) if compression is enabled on the server.
6. The server will then send a [Login Success Packet](https://wiki.vg/Protocol#Login_Success).
This is the server saying, "Okay, you are now logged in as this player".

7. The client now has to acknowledge this, and send a [Login Acknowledged Packet](https://wiki.vg/Protocol#Login_Acknowledged).
This is us saying, "Okay, now let us proceed to the configuration phase."

The server and client from this point will now send information about each other and server capabilities.

8. The server will finally send a [Finish Configuration Packet](https://wiki.vg/Protocol#Finish_Configuration),
9. Followed by the client sending an [Acknowledge Finish Configuration Packet](https://wiki.vg/Protocol#Acknowledge_Finish_Configuration).

Sending the Acknowledge Finish Configuration Packet will then move the server to the Play phase.

10. The server will now finally send the [Login (Play) Packet](https://wiki.vg/Protocol#Login_.28play.29), and now the server is ready to send information about the current game, such as chunk data/block data/entity data/world data.

At this point, you are free to send and receive any packets under the Play phase, however, you must ensure that you return the [Keep Alive Packet](https://wiki.vg/Protocol#Clientbound_Keep_Alive_.28play.29) when the server sends it in order to make sure that the server doesn't kick you out.

You may also refer to a different project called mineflayer, under https://github.com/PrismarineJS/mineflayer, which is a programmable headless (console) Minecraft client based on JavaScript.

It may help you understand some concepts a bit further when looking at it from a different client.
If you have any further questions, then you may look at various modding communities for help as well.

**Reply by iTsThermo** [reply]
Thank you for the quick response!!

**Comment by milutinke** [comment]
https://discord.gg/ZU2bdZ8N

A discord server for this exactly

**Reply by iTsThermo** [reply]
Thank you, I joined the server and I'm checking out the channels. I'm pretty excited to see what I can create.
